language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -155,43 +151,36 @@ typeCheckStep = do
 -- | Extract kind information from a type.
 kindOf :: Type TyNameWithKind a -> TypeCheckM a (Kind ())
 kindOf TyInt{} = pure (Size ())
-kindOf (TyFun x ty' ty'') = do
-    k <- kindOf ty'
-    k' <- kindOf ty''
-    if isType k && isType k'
-        then pure (Type ())
-        else
-            if isType k
-                then throwError (KindMismatch x (void ty'') k' (Type ()))
-                else throwError (KindMismatch x (void ty') k (Type ()))
+kindOf (TyFun x dom cod) = do
+    kindCheckM x dom $ Type ()
+    kindCheckM x cod $ Type ()
+    pure $ Type ()
 kindOf (TyForall x _ _ ty) = do
-    k <- kindOf ty
-    if isType k
-        then pure (Type ())
-        else throwError (KindMismatch x (void ty) (Type ()) k)
-kindOf (TyLam _ _ k ty) =
-    [ KindArrow () (void k) k' | k' <- kindOf ty ]
+    kindCheckM x ty $ Type ()
+    pure $ Type ()
+kindOf (TyLam _ _ argK body) = KindArrow () (void argK) <$> kindOf body
 kindOf (TyVar _ (TyNameWithKind (TyName (Name (_, k) _ _)))) = pure (void k)
 kindOf (TyBuiltin _ b) = do
     (TypeConfig _ (BuiltinTable tyst _)) <- ask
     case M.lookup b tyst of
         Just k -> pure k
         _      -> throwError InternalError
-kindOf (TyFix x _ ty) = do
-    k <- kindOf ty
-    if isType k
-        then pure (Type ())
-        else throwError (KindMismatch x (void ty) (Type ()) k)
-kindOf (TyApp x ty ty') = do
-    k <- kindOf ty
-    case k of
-        KindArrow _ k' k'' -> do
-            k''' <- kindOf ty'
+kindOf (TyFix x _ pat) = do
+    kindCheckM x pat $ Type ()
+    pure $ Type ()
+kindOf (TyApp x fun arg) = do
+    funK <- kindOf fun
+    case funK of
+        KindArrow _ argK resK -> do
             typeCheckStep
-            if k' == k'''
-                then pure k''
-                else throwError (KindMismatch x (void ty') k'' k''')
-        _ -> throwError (KindMismatch x (void ty') (KindArrow () (Type ()) (Type ())) k)
+            kindCheckM x arg argK
+            pure resK
+        _ -> throwError $ KindMismatch x (void fun) (KindArrow () dummyKind dummyKind) funK
+
+kindCheckM :: a -> Type TyNameWithKind a -> Kind () -> TypeCheckM a ()
+kindCheckM x ty k = do
+    tyK <- kindOf ty
+    when (tyK /= k) . throwError $ KindMismatch x (void ty) k tyK
```

Comment:
The `when p . throwError` might be simpler as `when p $ throwError` or something else.

Reply:
Agreed.